core/capabilities/gateway_connector/service_wrapper.go

```diff
@@ -0,0 +1,128 @@
+	gwConfig "github.com/smartcontractkit/chainlink/v2/core/config"
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	"github.com/smartcontractkit/chainlink/v2/core/services/gateway/connector"
+	"github.com/smartcontractkit/chainlink/v2/core/services/gateway/network"
+	"github.com/smartcontractkit/chainlink/v2/core/services/keystore"
+	"github.com/smartcontractkit/chainlink/v2/core/services/keystore/keys/ethkey"
+)
+
+type serviceWrapper struct {
+	services.StateMachine
+
+	config    *config.GatewayConnector
+	keystore  keystore.Eth
+	connector connector.GatewayConnector
+	lggr      logger.Logger
+}
+
+func translateConfigs(f config.GatewayConnector) connector.ConnectorConfig {
+	r := connector.ConnectorConfig{}
+	if f.NodeAddress != nil {
+		r.NodeAddress = f.NodeAddress()
+	}
+
+	if f.DonID != nil {
+		r.DonId = f.DonID()
+	}
+
+	if f.Gateways != nil {
+		r.Gateways = make([]connector.ConnectorGatewayConfig, len(r.Gateways))
+		for index, element := range f.Gateways() {
+			r.Gateways[index] = connector.ConnectorGatewayConfig{Id: element.ID(), URL: element.URL()}
+		}
+	}
+
+	if !reflect.ValueOf(f.WSHandshakeTimeoutMillis).IsZero() {
+		r.WsClientConfig = network.WebSocketClientConfig{HandshakeTimeoutMillis: f.WSHandshakeTimeoutMillis()}
+	}
+
+	if f.AuthMinChallengeLen != nil {
+		r.AuthMinChallengeLen = f.AuthMinChallengeLen()
+	}
+
+	if f.AuthTimestampToleranceSec != nil {
+		r.AuthTimestampToleranceSec = f.AuthTimestampToleranceSec()
+	}
+	return r
+}
+
+// NOTE: this wrapper is needed to make sure that our services are started after Keystore.
+func NewGatewayConnectorServiceWrapper(config *gwConfig.GatewayConnector, keystore keystore.Eth, lggr logger.Logger) *serviceWrapper {
+	return &serviceWrapper{
+		config:   config,
+		keystore: keystore,
+		lggr:     lggr,
+	}
+}
+
+func (e *serviceWrapper) Start(ctx context.Context) error {
+	return e.StartOnce("GatewayConnectorServiceWrapper", func() error {
+		conf := *e.config
+		e.lggr.Infow("Starting GatewayConnectorServiceWrapper", "chainID", conf.ChainIDForNodeKey())
+		chainId, _ := new(big.Int).SetString(conf.ChainIDForNodeKey(), 0)
+		enabledKeys, err := e.keystore.EnabledKeysForChain(ctx, chainId)
+		if err != nil {
+			return err
+		}
+		if len(enabledKeys) == 0 {
+			return errors.New("no available keys found")
+		}
+		configuredNodeAddress := common.HexToAddress(conf.NodeAddress())
+		idx := slices.IndexFunc(enabledKeys, func(key ethkey.KeyV2) bool { return key.Address == configuredNodeAddress })
+		if idx == -1 {
+			return errors.New("key for configured node address not found")
+		}
+		signerKey := enabledKeys[idx].ToEcdsaPrivKey()
+		if enabledKeys[idx].ID() != conf.NodeAddress() {
+			return errors.New("node address mismatch")
+		}
+		// cannot assign to conf.NodeAddress (neither addressable nor a map index expression
+		// I think I'm still confused about mixing interfaces and structures, and how to use a setter on an interface.
+		conf.NodeAddress = enabledKeys[idx].ID()
```

Comment:
You don't have to do this at all. You already used the configured address to find the needed key. No need to modify the config struct.